### PR TITLE
Add support for differentiating through prime and tag functions with keyword arguments

### PR DESCRIPTION
**Pull request by mtfishman**
Add support for differentiating through prime and tag functions with keyword arguments. An issue arises that certain combinations of arguments and keyword arguments make the prime or tag function difficult to invert. For example:
```julia
i = Index(2)
A = randomITensor(i', i)
prime(A; plev=1)
```
would require that the inverse is:
```julia
prime(A, -1; plev=2)
```
but the current functionality doesn't manipulate the keyword argument.

It is probably possible to keep track of each of these cases and do the correct manipulations in `inv_op`, but it is kind of nontrivial. For now I detect when the inverse operation is not working in the pullback by comparing to the indices of the ITensor input in the forward pass, and throw an error explaining that the user needs to write the prime or tag operation in an invertible way.

I also added some nontrivial examples involving energy minimization of an ITensor or an MPS, which for now don't use any AutoHOOT functionality (so just differentiate through pairwise contractions with non-optimal caching).

### CODE REVIEW

**Comment by LinjianMa** [comment]
@mtfishman Thanks for the updates, but I suggest let's not moving too fast on this, since there're corner cases still need to fix (mainly coming from higher-order derivatives). I suggest let's first get the AutoHOOT fully working, and then work on these things.

File: test/ITensorChainRules/runtests.jl [hunk 1]
```diff
@@ -4,6 +4,7 @@ using FiniteDifferences
 using ITensors
 using ITensors.NDTensors
 using ITensorNetworkAD
+using OptimKit
```
Comment (LinjianMa) [inline test/ITensorChainRules/runtests.jl:7]: Can separate this file into two, one testing chainrules another testing  energy minimization?

**Review by LinjianMa — COMMENTED** [review]
In the example given, plev is different for the forward pass and the backward pass. Is it implemented? I think currently the keyword argument is the same for both.

**Comment by mtfishman** [comment]
> In the example given, plev is different for the forward pass and the backward pass. Is it implemented? I think currently the keyword argument is the same for both.

Yes, I was explaining in my comment above that for now I don't implement that, but catch it and throw an error. It requires some thought to do it correctly in every case (especially when tags are involved) so I'll leave that for future work. In the case above you can do something like `replaceprime(A, 1 => 2)` which has an inverse of `replaceprime(A, 2 => 1)`, so usually you can rewrite it in a more obviously invertible way.

The main application is to do something like `prime(A; tags="x")` which only primes a subset of indices, which has an inverse of `prime(A, -1; tags="x")` (it doesn't require modifying the keyword arguments).

File: test/ITensorChainRules/runtests.jl (hunk 1 above)
Comment (mtfishman) [reply test/ITensorChainRules/runtests.jl:7]: Sure, makes sense.

**Comment by LinjianMa** [comment]
> > In the example given, plev is different for the forward pass and the backward pass. Is it implemented? I think currently the keyword argument is the same for both.
> 
> Yes, I was explaining in my comment above that for now I don't implement that, but catch it and throw an error. It requires some thought to do it correctly in every case (especially when tags are involved) so I'll leave that for future work. In the case above you can do something like `replaceprime(A, 1 => 2)` which has an inverse of `replaceprime(A, 2 => 1)`, so usually you can rewrite it in a more obviously invertible way.
> 
> The main application is to do something like `prime(A; tags="x")` which only primes a subset of indices, which has an inverse of `prime(A, -1; tags="x")` (it doesn't require modifying the keyword arguments).

I see, I didn't notice that you are testing the error using test_throws. Makes sense.

File: test/ITensorChainRules/runtests.jl (hunk 1 above)
Comment (mtfishman) [reply test/ITensorChainRules/runtests.jl]: I reorganized the tests in the latest commits.

**Comment by mtfishman** [comment]
This is good to go from my end, good to merge after tests pass?

**Comment by LinjianMa** [comment]
> This is good to go from my end, good to merge after tests pass?

Yes, once the tests pass we can merge this one.
